src/puppetlabs/puppetdb/command.clj

```diff
@@ -317,18 +433,124 @@
     (mig/activemq->stockpile config enqueue-fn dlo)
     (mig/lock-upgrade config)))
 
+;; The number of times a message can be retried before we discard it
+(def maximum-allowable-retries 5)
+
+(defn discard-message
+  "Discards the given `cmd` caused by `ex`"
+  [cmd ex q dlo]
+  (-> cmd
+      (queue/cons-attempt ex)
+      (dlo/discard-cmdref q dlo)))
+
+(defn process-delete-cmdref
+  "Processes a command ref marked for deletion. This is similar to
+  processing a non-delete cmdref except different metrics need to be
+  updated to indicate the difference in command"
+  [{:keys [command version] :as cmdref} q scf-write-db response-chan stats]
+  (process-command-and-respond! cmdref scf-write-db response-chan stats)
+  (queue/ack-command q {:entry (queue/cmdref->entry cmdref)})
+  (update-counter! :depth command version dec!)
+  (update-counter! :invalidated command version dec!))
+
+(defn process-cmdref
+  "Parses, processes and acknowledges a successful command ref and
+  updates the relevant metrics. Any exceptions that arise are
+  unhandled and expected to be caught by the caller."
+  [cmdref q scf-write-db response-chan stats]
+  (let [{:keys [command version] :as cmd} (queue/cmdref->cmd q cmdref)
+        retries (count (:attempts cmdref))]
+
+    (create-metrics-for-command! command version)
+
+    (mark-both-metrics! command version :seen)
+    (update! (global-metric :retry-counts) retries)
+    (update! (cmd-metric command version :retry-counts) retries)
```

Comment:
Might be exactly what we want, but just noticed that we bump the retry counts before processing (suppose if we didn't want that, the alternative would be in a finally block or similar).